File: 2090-number-of-ways-to-arrive-at-destination/number-of-ways-to-arrive-at-destination.py
class Solution:
    def graph(self, roads):
        adj={}
        for u,v,i in roads:
            if u not in adj:
                adj[u] = []
            if v not in adj:
                adj[v] = []
            adj[u].append((v,i))
            adj[v].append((u,i))
        return adj

    def shortestPath(self, V , adj , src):
        heap=[]
        dist=[float('inf')]*V
        dist[src]=0
        ways=[0]*V
        ways[src]=1
        heapq.heappush(heap,(0,src))
        while heap:
            distance , u = heapq.heappop(heap)
            if distance > dist[u]:
                continue
            for dest,weight in adj.get(u,[]):
                if weight+dist[u]< dist[dest]:
                    dist[dest]=weight+dist[u]
                    ways[dest] = ways[u]
                    heapq.heappush(heap,(dist[dest],dest))
                   
                elif (dist[u]+weight)==dist[dest]:
                    ways[dest]=(ways[dest]+ways[u])%(10**9 + 7)


        return ways[V-1]

    def countPaths(self, n: int, roads: List[List[int]]) -> int:

        g=self.graph(roads) 
        return self.shortestPath(n,g,0)





        

File: 2090-number-of-ways-to-arrive-at-destination/number-of-ways-to-arrive-at-destination.java
class Solution {
    public static class Pair{
        int src;
        int dest;
        long time;
        Pair(int src , int dest , long time){
            this.src=src;
            this.dest=dest;
            this.time=time;
        }
          Pair(int src , long time){
            this.src=src;
            this.time=time;
        }
    }

    public HashMap<Integer, List<Pair>> createGraph(int n, int[][] roads ){
        HashMap<Integer, List<Pair>> map = new HashMap<>();
        for(int i=0;i<n;i++){
            map.put(i,new ArrayList<>());
        }
        for(int[] road : roads){
            map.get(road[0]).add(new Pair(road[0],road[1],road[2]));
            map.get(road[1]).add(new Pair(road[1],road[0],road[2]));
        }
        return map;
    }
    public int countPaths(int n, int[][] roads) {
        PriorityQueue<Pair> queue = new PriorityQueue<>(Comparator.comparingLong(a->a.time));
        HashMap<Integer,List<Pair>> adj = createGraph(n,roads);
        long[] distance = new long[n];
        long[] ways = new long[n];
        Arrays.fill(distance,Long.MAX_VALUE);
        ways[0]=1;
        distance[0]=0;
        queue.add(new Pair(0,0));
        while(!queue.isEmpty()){
            Pair curr = queue.poll();
            if (curr.time > distance[curr.src]) {
    continue;
}
            for(int i=0;i< adj.get(curr.src).size();i++){

                Pair node = adj.get(curr.src).get(i);
                if(distance[node.dest]> node.time + distance[node.src]){
                    distance[node.dest]=node.time+distance[node.src];
                    ways[node.dest]=ways[node.src];
                    queue.add(new Pair(node.dest,distance[node.dest]));
                }
                else if( distance[node.dest]== node.time + distance[node.src]){
                    ways[node.dest]=(ways[node.dest]+ways[node.src])%1_000_000_007L;
                }

            }
        }

      return (int)ways[n-1];
    }
}